ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -49,56 +69,80 @@ private[kvutils] object InputsAndEffects {
       coid match {
         case acoid: AbsoluteContractId =>
-          InsertOrdSet.empty + absoluteContractIdToStateKey(acoid)
+          inputs += absoluteContractIdToStateKey(acoid)
         case _ =>
-          InsertOrdSet.empty
       }
 
-    def partyInputs(parties: Set[Party]): InsertOrdSet[DamlStateKey] = {
+    def partyInputs(parties: Set[Party]): List[DamlStateKey] = {
       import Party.ordering
-      InsertOrdSet.fromSeq(parties.toList.sorted.map(partyStateKey))
+      parties.toList.sorted.map(partyStateKey)
     }
 
-    tx.fold(packageInputs: Set[DamlStateKey]) {
-        case (inputs, (nodeId, node)) =>
-          node match {
-            case fetch: NodeFetch[ContractId] =>
-              inputs ++ contractInputs(fetch.coid) ++ partyInputs(fetch.signatories) ++
-                partyInputs(fetch.stakeholders)
-            case create: NodeCreate[ContractId, VersionedValue[ContractId]] =>
-              create.key.fold(inputs) { keyWithM =>
-                inputs + contractKeyToStateKey(
-                  GlobalKey(create.coinst.template, forceNoContractIds(keyWithM.key)))
-              } ++ partyInputs(create.signatories) ++ partyInputs(create.stakeholders)
-            case exe: NodeExercises[_, ContractId, _] =>
-              inputs ++ contractInputs(exe.targetCoid) ++ partyInputs(exe.stakeholders) ++ partyInputs(
-                exe.signatories) ++ partyInputs(exe.controllers)
-            case l: NodeLookupByKey[ContractId, Transaction.Value[ContractId]] =>
-              // We need both the contract key state and the contract state. The latter is used to verify
-              // that the submitter can access the contract.
-              l.result.fold(inputs)(inputs ++ contractInputs(_)) +
-                contractKeyToStateKey(GlobalKey(l.templateId, forceNoContractIds(l.key.key)))
-          }
-      }
-      .toList
+    tx.foreach {
+      case (_, n) =>
+        n match {
+          case fetch: NodeFetch[ContractId] =>
+            addContractInput(fetch.coid)
+
+          case create: NodeCreate[ContractId, VersionedValue[ContractId]] =>
+            create.key.foreach { keyWithM =>
+              inputs += contractKeyToStateKey(
+                GlobalKey(create.coinst.template, forceNoContractIds(keyWithM.key)))
+            }
+
+          case exe: NodeExercises[_, ContractId, _] =>
+            addContractInput(exe.targetCoid)
+
+          case l: NodeLookupByKey[ContractId, Transaction.Value[ContractId]] =>
+            // We need both the contract key state and the contract state. The latter is used to verify
+            // that the submitter can access the contract.
+            l.result.foreach(addContractInput)
+            inputs += contractKeyToStateKey(GlobalKey(l.templateId, forceNoContractIds(l.key.key)))
+        }
+
+        inputs ++= partyInputs(n.informeesOfNode)
+    }
+
+    inputs.toList
   }
 
+  def subValues(v: Value[ContractId]): List[Value[ContractId]] =
```

Comment:
Where is this actually used? Can't find a reference to it anywhere.

Reply:
@SamirTalwar-DA it's being reworked (unfortunately can't move it back to draft)